### ISSUE TITLE
[Question]: Recovering from a MsalThrottledUiRequiredException.

### ISSUE THREAD

**Issue by SunshineCoast1000** [AzureAD/microsoft-authentication-library-for-dotnet#2256]
I'm not sure if this is a bug or a change made to Microsoft.Identity.Client (M.I.C) between version 4.8.1 and 4.23.0.

But I used to be able to 'reset' a silent signin as below ...
{note: this isn't exactly my coding but it should explains my problem)
----------------------------------------------------------------------------

```CSharp
GraphClient = new GraphServiceClient(new DelegateAuthenticationProvider(async (request) =>
{
	try //silent signin
	{
		var silent = await PCA.AcquireTokenSilent(Scopes, Accounts.First()).ExecuteAsync();
		//ABOVE FAILS WITH MsalUiRequiredException on both M.I.C version 4.8.1 and 4.23.0

		IsSilent = true; //flag silent is good
	}
	catch (MsalUiRequiredException ex)
	{
		if (IsSilent) //try silent reset
		{
			GraphClient = new GraphServiceClient(new DelegateAuthenticationProvider(async (request2) =>
			{
				var silent = await PCA.AcquireTokenSilent(Scopes, Accounts.First()).ExecuteAsync();
				//ABOVE USED TO BE OK WITH M.I.C version 4.8.1 AND I COULD CONTINUE BROWSING ONEDRIVE
				//BUT IT NOW FAILS WITH A MsalThrottledUiRequiredException using M.I.C version 4.23.0
			}));
		}
		else
		{
			//user interactive sign in here ...
		}
	}
}));
```

----------------------------------------------------------------------------
Using Microsoft.Graph 			Version 1.21.0
Using Microsoft.Identity.Client 	Version 4.8.1

These exceptions are thrown, 

Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
Exception thrown: 'Microsoft.Graph.ServiceException' in System.Private.CoreLib.dll

However,

Using Microsoft.Graph 			Version 3.20.0
Using Microsoft.Identity.Client 	Version 4.23.0

These exceptions are thrown, 

Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
Exception thrown: 'Microsoft.Identity.Client.MsalThrottledUiRequiredException' in System.Private.CoreLib.dll
Exception thrown: 'Microsoft.Graph.ServiceException' in System.Private.CoreLib.dll
----------------------------------------------------------------------------

Has Microsoft.Identity.Client made a decisive change to completely disable looping 
clients and have no option for recovery other than an interactive signin, 
or is this a problem with an M.I.C upgrade ?

**Comment by jmprieur** [comment]
@SunshineCoast1000 : given MsalThrottledUiRequiredException inherits from MsalUiRequiredException don't you think that your catch(MsalUiRequiredException) was in fact already catching an MsalThrottledUiRequiredException ?

**Comment by jmprieur** [comment]
Also I don't understand how resetting the Graph DelegateAuthenticationProvider ever worked, given this is the same confidential client application.

**Comment by SunshineCoast1000** [comment]
I apologize for the horrendous formatting as I'm a first time user on GitHub.

The code in my message isn't exact as I am trying to reduce it some but basically I can reset a silent sign in
using M.I.C  4.8.1 but now I have upgraded to M.I.C 4.23.0 I can't do it anymore.

I suspect that MsalThrottledUiRequiredException has been added to an upgrade version of M.I.C 4.8.1
and it has stopped me from reseting a silent signin.

**Comment by bgavrilMS** [comment]
I'm not sure what you mean by "reset a silent login"? 

We have introduced a throttling feature to prevent apps from calling `AcquireTokenSilent` when we know it would fail. In you code sample, you make 2 calls to `AcquireTokenSilent`. If the first call fails because ESTS tells you "you need to interactievly login this user" (e.g. consent is required, or 2FA etc.), then there is no point in calling `AcquireTokenSilent` again with the same params because you'd get the same result. So MSAL blocks you to remove some load from the service.

As @jmprieur pointed out, MsalUiThrottledException is a subtype of MsalUiRequiredException, so the catch block should continue to function. 

From the code you pasted above, it is not clear when you make first `AcquireTokenSilent` call that fails and when you make the second call. Generally we recommend the pattern

```
try
   AcquireTokenSilent
catch UiRequiredException
   AcquireTokenInteractive
```

**Comment by SunshineCoast1000** [comment]
I apologize for the confusion but its difficult as an independent developer to know how to do everything.  My 'reseting' a silent signin works with M.I.C 4.8.1 so I will keep using it and we can close this thread.  Thanx for your time and knowledge.

**Comment by bgavrilMS** [comment]
Would it be possible to get a network trace (e.g. using Fiddler) trace to see what's going on? We fix many things in subsequent versions of MSAL, staying on an old version is not great. We may have a bug in the throttling logic, and would definitely like to understand if that is the case. 

The Fiddler trace may contain sensitive information, so please send it via email at bogavril @ microsoft.com

**Comment by bgavrilMS** [comment]
Or logging, as described here: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/logging#logging-in-msal

**Comment by SunshineCoast1000** [comment]
Using M.I.C 4.8.1 I do this,

1. Create a PublicClientApplicationBuilder > Signin > GraphServiceClient > browse files on OneDrive 
2. When MsalUiRequiredException fired due to looping I basically did a 'reset' by doing the above again
3. RESET > Create a new PublicClientApplicationBuilder > GraphServiceClient > browse files on OneDrive 

but with M.I.C 4.23.0 I can't do that anymore because this MsalThrottledUiRequiredException has been introduced.

**Comment by bgavrilMS** [comment]
At step 2, can you detail what this MsalUiRequiedException looked like? Is there an error code like "invalid_grant" ? What was the error message?

**Comment by SunshineCoast1000** [comment]
with M.I.C 4.8.1 I get this,
> GetFiles from OneDrive Folder ...
> GetFiles from OneDrive Folder ...
> GetFiles from OneDrive Folder ...

Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
OpenGraphClient: MsalException = AADSTS50196: The server terminated an operation because it encountered a client request loop. Please contact your app vendor.
Trace ID: d35bdd8d-2aaf-4eb2-a466-a3d68ccbd101
Correlation ID: 72b321c8-23fd-4b6a-b663-bf9e851d9498
Timestamp: 2020-12-01 14:23:06Z
OpenGraphClient: IsGraphClient = True

Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
OpenGraphClient: MsalException = AADSTS50196: The server terminated an operation because it encountered a client request loop. Please contact your app vendor.
Trace ID: d35bdd8d-2aaf-4eb2-a466-a3d690cbd101
Correlation ID: d0b9321e-57f5-4364-83d0-ebe7c9445c9f
Timestamp: 2020-12-01 14:23:06Z
OpenGraphClient: IsGraphClient = True

Exception thrown: 'Microsoft.Graph.ServiceException' in System.Private.CoreLib.dll

GetFiles RESET
> GetFiles from OneDrive Folder ...
> GetFiles from OneDrive Folder ...
> GetFiles from OneDrive Folder ...

Basically I can just continue browsing OneDrive forever with my GetFiles RESET

but with M.I.C 4.23.0 I get this,
> GetFiles from OneDrive Folder ...
> GetFiles from OneDrive Folder ...
> GetFiles from OneDrive Folder ...

Exception thrown: 'Microsoft.Identity.Client.MsalUiRequiredException' in System.Private.CoreLib.dll
Exception thrown: 'Microsoft.Identity.Client.MsalThrottledUiRequiredException' in System.Private.CoreLib.dll
Exception thrown: 'Microsoft.Graph.ServiceException' in System.Private.CoreLib.dll

and then I can't RESET

**Comment by SunshineCoast1000** [comment]
I believe that this new MsalThrottledUiRequiredException terminates silent signin capability and forces an interactive signin ... but I also suspect that even after you do the forced interactive signin, you can't get a silent signin again because the MsalThrottledUiRequiredException keeps forcing an interactive signin continuously.

**Comment by bgavrilMS** [comment]
Ok, this is starting to make more sense. ESTS (the server) is blocking you because you request tokens too often "the server terminated an operation because it encountered a client request loop". "Request Loop" is a what ESTS calls a throttling event, i.e. you make the same request too often.

With MSAL 4.8.1, simply catching the exception and making a new request is probably enough time (a few ms!) to "reset the loop", i.e. ESTS no longer throttles you. 
With MSAL 4.23.0, the library itself is detecting the loop and it much more strict, it will block you from making the exact same silent request for a whole 2 minutes.


Now let's move on to the real root cause of this issue. You correctly call `AcquireTokenSilent`, which is supposed to give you a token from MSAL's cache. Tokens are valid for 1h after all. But somehow the token cache isn't hit and MSAL makes a server request to ESTS. There must be something wrong with MSAL's token cache logic or with the scopes. 

Could you tell me what scopes you are requesting?

**Comment by SunshineCoast1000** [comment]
I'm still in product development but these are the scopes i been testing with,

Scopes = "User.ReadWrite.All offline_access Directory.ReadWrite.All offline_access Files.ReadWrite.All offline_access Sites.ReadWrite.All offline_access";

**Comment by bgavrilMS** [comment]
I will have a look at this, it's probably a bug in MSAL (thanks for bearing with me!). 

For now, delete `offline_access`. MSAL adds `offline_access` for you anyway, it is required for its inner logic. Hopefully this will solve the issue.

**Comment by SunshineCoast1000** [comment]
OK ... many thanx for your help and time ... I should be fine with my work as I generally find a work around for everything.

**Comment by bgavrilMS** [comment]
No problem, and thanks for reporting this @SunshineCoast1000 . It's a fairly important bug. 
Let me know if things work without `offline_access` if you get the change to try it out.

**Comment by SunshineCoast1000** [comment]
My problem is resolved and working fine ... however ... just as a quick observation,

A 'File Picker' must allow the user to browse a directory all they want to open/save a folder/file.  Because mobile platforms are now restricting external storage, mobile apps need the ability to open/save data on cloud accounts. Therefore we need the ability to produce quality 'File Pickers' for cloud accounts that don't have restrictions on looping.

ie.  you can't say to users that your only allowed to browse folders up to 20 times in 2 mins for example. 

A solution could be to provide a method in Microsoft.Identity.Client that retrieves the complete folders/files collection starting from the root directory in one request call.  That way mobile developers can produce quality File Pickers that only make one request call on an account and allow the user to browse the retrieved folder/file collection all they want from then on.

**Comment by bgavrilMS** [comment]
Microsoft.Identity.Client is only responsible for getting you a token for the scopes you request. I agree with you that we should minimize the number of API calls, but I do not know the inner workings of Graph. 

The Graph service and SDK are responsible for what you describe, i.e. better APIs around cloud storage, so I suggest you open a issue with the Graph SDK or you open a case in the Azure Portal around this.

**Comment by jmprieur** [comment]
And the graph SDK is a wrapper around the graph API.
